### DISCUSSION TITLE
Looking for resources about SRD

### DISCUSSION THREAD

**Discussion by simibac** [decentralized-identity/veramo#356] [Q&A]
Since the workflow of a Selective Resource Disclosure (SRD) request and response is not fully clear to me, I was wondering what resources were used to build the Veramo SRD requests and responses. 

I have read the official [documentation about SRD](https://www.w3.org/TR/vc-data-model/#concrete-lifecycle-example) of W3C but I do not understand entirely what the workflow looks like.
1. I don't understand how and where "challenge" and "domain" in the [verifiable presentation ](https://www.w3.org/TR/vc-data-model/#example-2-a-simple-example-of-a-verifiable-presentation)are generaded.
2. When I use the veramo cli and try to create an SRD response it `veramo sdr respond`, it does not show any options to choose from and results in the following exception:

```
$veramo sdr respond                                                   
? Selective disclosure request (Use arrow keys)

/usr/local/lib/node_modules/@veramo/cli/node_modules/inquirer/lib/prompts/list.js:148
    return this.opt.choices.getChoice(this.selected).value;
                                                    ^

TypeError: Cannot read property 'value' of undefined
    at ListPrompt.getCurrentValue (/usr/local/lib/node_modules/@veramo/cli/node_modules/inquirer/lib/prompts/list.js:148:53)
    at MapSubscriber._next (/usr/local/lib/node_modules/@veramo/cli/node_modules/rxjs/internal/operators/map.js:49:35)
    at MapSubscriber.Subscriber.next (/usr/local/lib/node_modules/@veramo/cli/node_modules/rxjs/internal/Subscriber.js:66:18)
    at TakeSubscriber._next (/usr/local/lib/node_modules/@veramo/cli/node_modules/rxjs/internal/operators/take.js:54:30)
    at TakeSubscriber.Subscriber.next (/usr/local/lib/node_modules/@veramo/cli/node_modules/rxjs/internal/Subscriber.js:66:18)
    at Interface.handler (/usr/local/lib/node_modules/@veramo/cli/node_modules/rxjs/internal/observable/fromEvent.js:22:28)
    at Interface.emit (node:events:388:22)
    at Interface._onLine (node:readline:363:10)
    at Interface._line (node:readline:699:8)
    at Interface._ttyWrite (node:readline:1044:14)
```
How can I create a SRD response? Is there some documentation that shows the entire workflow of SRD?

3. I was wondering if anybody can point me towards a more detailed example/documentation with ZKP in combo with verifiable credentials. I read [this](https://www.w3.org/TR/vc-data-model/#zero-knowledge-proofs) but I would like to learn more about derived verifiable credential and ZNP. Also if there is a working prototype, that would be extremely helpful.

Appreciate any help!

**Comment by mirceanis** [comment]
Selective disclosure support in veramo is still in early stages, with regards to W3C alignment.
The current implementation is heavily inspired by the uPort [selective disclosure protocol](https://github.com/uport-project/specs/blob/develop/flows/selectivedisclosure.md), but uses a simpler approach.

There are a few key differences between the W3C discussions and the uPort SDR flow:
* uPort did not require `domain` and `challenge` parameters since the original request is embedded in the response.
* uPort made use of atomic credentials (one or very few claims per credential), reducing the need for ZKP or derived credentials
* uPort was using only JWT encoding

We are aware that the approach contained in this repository does not align to community agreed specs and are working to address this.
This is also the reason that the current implementation of SDR is not very well documented yet.

Throughout the codebase here, there is still an assumption that atomic credentials are in use.
But this is subject to change since the plan is to adopt a broader approach to credential and proof formats.
The main effort now is to add support for JSON-LD credentials and presentations.
This should greatly improve alignment since most of the W3C spec refers to the LD format.
The next step is to make some improvements to the verification API so that checks for `domain` and `challenge` can be performed seamlessly.
We also plan to support BBS+ in the next few months which would allow derived credentials in selective disclosure responses.

@simonas-notcat @jasonhealy can you perhaps point to an example of the existing SDR flow using the veramo CLI?

**Reply by simibac** [reply]
> uPort did not require domain and challenge parameters since the original request is embedded in the response.

By looking at the docs, it is not clear to me where the request is embedded in the response. Either this is not a complete response or I am missing something. What is it? :)

Also, I assume that the `tag` does not have to be listed in the VP since it is part of the SDR. And I can get it from there. But how is the SDR part of the VP?

This is taken from the open-api.json for `/createVerifiablePresentation`:
```js
{
  "presentation": {
    "id": "string",
    "holder": "string",
    "issuanceDate": "string",
    "expirationDate": "string",
    "@context": [
      "string"
    ],
    "type": [
      "string"
    ],
    "verifier": [
      "string"
    ],
    "verifiableCredential": [
      {
        "@context": [
          "string"
        ],
        "id": "string",
        "type": [
          "string"
        ],
        "issuer": {
          "id": "string"
        },
        "issuanceDate": "string",
        "expirationDate": "string",
        "credentialSubject": {
          "id": "string"
        },
        "credentialStatus": {
          "id": "string",
          "type": "string"
        },
        "proof": {
          "type": "string"
        }
      }
    ]
  },
  "save": true,
  "proofFormat": "string"
}
```

**Reply by mirceanis** [reply]
You are not missing something, what you are looking for is not there (yet).

I think I misrepresented the current implementation of `selective-disclosure` with my answer, and I apologise if I wasn't clear.
The current implementation is inspired by the uPort selective disclosure but doesn't have all the features from it. It doesn't include the original request in the response, like the uPort app did.
We are aware that there are other efforts to define protocols to exchange credentials and presentations and did not want to put too much effort in [defining a new one](https://xkcd.com/927/) since the plan is to make use of more community standards.
One of the goals of Veramo is to be a framework where protocols like these can be developed and exercised.

Our development resources are limited so there was not enough time to implement everything yet.
We're doing our best to provide a more complete stack, but things will obviously take some time.

We appreciate all contributions, so if something is missing, thank you for signalling it! We will bump it up in the priority list.
If you describe it in enough detail, maybe another community member will pick it up and provide fix or an implementation ;)

**Reply by simibac** [reply]
Thanks for clarifying! I appreciate all the work you do. It is extremely helpful.
